### ISSUE TITLE
MKLDNN convolution leaks memory

### ISSUE THREAD

**Issue by hrpan** [pytorch/pytorch#21537]
## 🐛 Bug

Memory usage increasing steadily when repeatedly calling forward.

## To Reproduce

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class Net(torch.jit.ScriptModule):

    def __init__(self):
        super(Net, self).__init__()

        self.conv = nn.Conv2d(1, 1, 1)

        self.lin = nn.Linear(100, 1)

    @torch.jit.script_method
    def forward(self, x):
        x = F.relu(self.conv(x))
        x = x.view(x.shape[0], -1)
        x = F.relu(self.lin(x))

        return x

m = Net()

x = torch.ones((1, 1, 10, 10))

for i in range(1000000):
    m(x)
```

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-16ubuntu3) 7.3.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.14.2
[pip3] numpydoc==0.7.0
[pip3] torch==1.1.0
[pip3] torchvision==0.3.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2018.0.1             h19d6760_4
[conda] mkl-dnn                   0.14                          2    intel
[conda] mkl-service               1.1.2            py36h17a0993_4
[conda] pytorch-cpu               1.1.0               py3.6_cpu_0    pytorch
[conda] torchvision-cpu           0.3.0             py36_cuNone_1    pytorch

## Additional context

Used memory_profiler to record the memory usage by `mprof run script.py`. Results are shown below.
![leak](https://user-images.githubusercontent.com/18377357/59131260-790c4580-89a4-11e9-8469-c3f8f12a772b.png)

**Comment by zdevito** [comment]
@hrpan Thanks for the code, we will look into the leak.

**Comment by zdevito** [comment]
A minimal version of the code that still leaks is the following. It no longer uses the jit, so it looks like the issue is in the F.conv2d function. I am removing the JIT tag because it does not appear related.

```
import torch
import torch.nn.functional as F
import time

weight = torch.ones([1,1,1,1])
x = torch.ones((1, 1, 10, 10))

oo = (1, 1)
zz = (0, 0)

start = time.time()
with torch.no_grad():
    while time.time() - start < 10:
        F.conv2d(x, weight, None, oo, zz, oo, 1)
```

**Comment by zdevito** [comment]
More investigation indicates that this is a leak in the mkldnn convolution pathway. In particular, disabling mkldnn by disabling the if statement on `Convolution.cpp:445` makes the leak disappear. Benchmarking script is:
```
pip install mprof
mprof run script.py && mprof plot -o foo.png 
```

**Comment by bddppq** [comment]
@Jianhui-Li @XiaobingSuper @gujinghui @mingfeima @jgong5 Could you help investigate?

**Comment by jgong5** [comment]
@uyongw

**Comment by jgong5** [comment]
@bddppq We will follow up.

**Comment by houseroad** [comment]
Some of our internal training jobs are also affected by this memory leak. After disabling the mkldnn, the problem is gone. Could you @jgong5 prioritize this issue? Thanks in advance.

**Comment by yinghai** [comment]
@jgong5 Our profiling shows that this place is holding all the memory 

https://github.com/intel/ideep/blob/pytorch/include/ideep/tensor.hpp#L513

**Comment by houseroad** [comment]
Some context from my profiling:
Stack
https://github.com/intel/ideep/blob/d7304e0345c3f0647cb020a71682d680b9f0424a/include/ideep/computations.hpp#L595
=>
https://github.com/intel/ideep/blob/d7304e0345c3f0647cb020a71682d680b9f0424a/include/ideep/computations.hpp#L539

**Comment by Jianhui-Li** [comment]
@yinghai @houseroad  Thanks for the info. 

@uyongw  Any update?

**Comment by gujinghui** [comment]
we're working on this already. will give update ASAP.

**Comment by uyongw** [comment]
Thanks for reporting this issue. We've been able to identify the issue and come up with a workaround as below. A formal fix will be submitted soon. 

A simple workaround in mkldnn src/common/stream.cpp.
https://github.com/intel/mkl-dnn/blob/rls-v0.18/src/common/stream.cpp

diff --git a/src/common/stream.cpp b/src/common/stream.cpp
index 054fbb9..be11cf0 100644
--- a/src/common/stream.cpp
+++ b/src/common/stream.cpp
@@ -46,7 +46,9 @@ status_t stream_t::submit(const nstl::vector<primitive_t *> &prims,

     const size_t start = stream_.size();
     stream_.insert(stream_.end(), prims.begin(), prims.end());
-    return submit_impl(start, stream_.size(), error_prim);
+    auto res = submit_impl(start, stream_.size(), error_prim);
+    stream_.clear();
+    return res;
}

**Comment by gujinghui** [comment]
Fixed by #22392
